ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java

```diff
@@ -45,33 +62,213 @@ public void migrate() {
         AtomicInteger updatedClaims = new AtomicInteger(0);
         AtomicInteger failedMigrations = new AtomicInteger(0);
 
-        notMigratedClaims.forEach(claim -> {
-            try {
-                logger.info("\t\t start migrating claim: " + claim.getReferenceNumber());
+        notMigratedClaims.sort(Comparator.comparing(Claim::getId));
 
-                Optional<Long> ccdId = coreCaseDataService.getCcdIdByReferenceNumber(user, claim.getReferenceNumber());
-                if (ccdId.isPresent()) {
-                    coreCaseDataService.overwrite(user, ccdId.get(), claim);
-                    logger.info("\t\t claim exists - overwrite");
-                    updatedClaims.incrementAndGet();
-                } else {
-                    coreCaseDataService.create(user, claim);
-                    logger.info("\t\t claim created in ccd");
-                    migratedClaims.incrementAndGet();
-                }
-
-                claimRepository.markAsMigrated(claim.getId());
+        notMigratedClaims.forEach(claim -> {
+            Optional<CaseDetails> caseDetails
+                = coreCaseDataService.getCcdIdByReferenceNumber(user, claim.getReferenceNumber());
 
-                logger.info("\t\t migrated successfully claim: " + claim.getReferenceNumber());
-            } catch (Exception e) {
-                logger.info(e.getMessage(), e);
-                failedMigrations.incrementAndGet();
+            if (!caseDetails.isPresent()) {
+                createCase(user, migratedClaims, claim, CaseEvent.SUBMIT_PRE_PAYMENT, failedMigrations);
+                updateCase(user, updatedClaims, failedMigrations, claim, null);
+            } else {
+                updateCase(user, updatedClaims, failedMigrations, claim, caseDetails.get());
             }
+
         });
 
         logger.info("Total Claims in database: " + notMigratedClaims.size());
         logger.info("Successfully migrated: " + migratedClaims.toString());
         logger.info("Successfully updated: " + updatedClaims.toString());
         logger.info("Failed to migrate: " + failedMigrations.toString());
     }
+
+    private void updateCase(
+        User user,
+        AtomicInteger updatedClaims,
+        AtomicInteger failedMigrations,
+        Claim claim,
+        CaseDetails caseDetails
+    ) {
+
+        CaseDetails ccdCase = Optional.ofNullable(caseDetails).orElseGet(() ->
+            coreCaseDataService.getCcdIdByReferenceNumber(user, claim.getReferenceNumber()).orElse(null));
+
+        if (ccdCase == null) {
+            return;
+        }
+
+        try {
+            for (CaseEvent event : CaseEvent.values()) {
```

Comment:
Prefer not to rely on the order of enums to decide what event to fire next. I think we should be able to decide based on the current state of claim in CCD.

Reply:
can test later with different implementation, but we don't have timestamp for all event, so it is tough to find the sequence

Reply:
Ok, we can defer that and will implement if there is any issue found in dry run.